### PR TITLE
feat(approve): adjust partial approves for buy orders

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes https://www.notion.so/cownation/Handle-edge-case-for-Buy-orders-with-EOAs-2578da5f04ca8083a97de4eb393b4ac5
https://www.notion.so/cownation/Wrong-navigation-when-approval-TX-fails-2878da5f04ca801982fede0f21a79bc7
https://www.notion.so/cownation/Don-t-navigate-to-confirm-moda-when-approved-less-than-needed-2878da5f04ca80028f02e0ca16445316

1. Added "Current allowance" displaying when approval is needed ([commit](https://github.com/cowprotocol/cowswap/pull/6376/commits/a641192693556bd3f9f88415ad3561e64595c780))
2. Added 1% on top of the amount to approve for buy orders ([commit](https://github.com/cowprotocol/cowswap/pull/6376/commits/a641192693556bd3f9f88415ad3561e64595c780))
3. Added "optimistic" allowance updating based on approval tx receipt. I added that to immediately update the allowance when tx is mined ([commit](https://github.com/cowprotocol/cowswap/pull/6376/commits/f583a9bb1fd3347123f7d4ae1803f7db680c9240))
4. Fixed navigation to confirm screen after swap. Before we could go there when: a) tx was not successfully mined. b) user has lowered the allowance on the wallet side. Now, in those cases we don't go to confirm screen. See the demo video ([commit](https://github.com/cowprotocol/cowswap/pull/6376/commits/62f2680376653059bad2ddc0588f3c894be08e37))

Screenshot 2025-10-16 at 15 28 21

Screenshot 2025-10-16 at 15 50 18

https://github.com/user-attachments/assets/d309eb09-f62d-4327-b25d-8c5066225b20

# To Test

1. https://www.notion.so/cownation/Handle-edge-case-for-Buy-orders-with-EOAs-2578da5f04ca8083a97de4eb393b4ac5
2. https://www.notion.so/cownation/Wrong-navigation-when-approval-TX-fails-2878da5f04ca801982fede0f21a79bc7
3. https://www.notion.so/cownation/Don-t-navigate-to-confirm-moda-when-approved-less-than-needed-2878da5f04ca80028f02e0ca16445316
4. See the cases above

## Summary by CodeRabbit

* **New Features**
  * Global optimistic allowance tracking with automatic expiry and network reset.
  * Current token allowance shown near approval controls; consolidated approve-and-swap flow with clearer approved-amount results.
  * Permission tooltip only shown when no cached permit; confirm button text adapts for bridging.

* **Bug Fixes**
  * UI uses actual on-chain approval amounts when wallet edits occur.
  * Reduced approval/swap race and error scenarios; standardized approval error handling and analytics.

* **Enhancements**
  * Buy orders get a 1% send-amount buffer for better fills.

* **Tests**
  * Extensive coverage added for approval, permit, and approval-result flows.

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/trade/hooks/useAmountsToSignFromQuote.ts [hunk 1]
```diff
@@ -16,14 +19,20 @@ export interface AmountsToSign {
...
   return useMemo(() => {
     const maximumSendSellAmount = isQuoteBasedOrder ? afterSlippage?.sellAmount : inputCurrencyAmount
     const minimumReceiveBuyAmount = isQuoteBasedOrder ? afterSlippage?.buyAmount : outputCurrencyAmount
 
     if (!maximumSendSellAmount || !minimumReceiveBuyAmount) return null
 
-    return { maximumSendSellAmount, minimumReceiveBuyAmount }
-  }, [isQuoteBasedOrder, inputCurrencyAmount, outputCurrencyAmount, afterSlippage])
+    return {
+      // Add 1% threshold for buy orders to level out price/gas fluctuations
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/trade/hooks/useAmountsToSignFromQuote.ts:31]: 1% addition to approving amount for buy orders

File: apps/cowswap-frontend/src/modules/onchainTransactions/updaters/FinalizeTxUpdater/services/finalizeEthereumTransaction.ts [hunk 2]
```diff
@@ -16,21 +16,13 @@ export function finalizeEthereumTransaction(
...
-  const { chainId, dispatch, updateLastApproveTxBlockNumber } = params
+  const { chainId, dispatch } = params
   const { hash } = transaction
 
   console.log(`[FinalizeTxUpdater] Transaction ${receipt.transactionHash} has been mined`, receipt, transaction)
 
   ONCHAIN_TRANSACTIONS_EVENTS.emit(OnchainTxEvents.BEFORE_TX_FINALIZE, { transaction, receipt })
 
-  // Once approval tx is mined, we add the priority allowance to immediately allow the user to place orders
-  if (transaction.approval) {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/onchainTransactions/updaters/FinalizeTxUpdater/services/finalizeEthereumTransaction.ts:27]: It was removed because now we have optimistic allowance

File: apps/cowswap-frontend/src/modules/erc20Approve/hooks/useApproveAndSwap.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    const amountToApproveBig = BigInt(amountToApprove.quotient.toString())
+    const toApprove = isPartialApproveEnabledByUser ? amountToApproveBig : MAX_APPROVE_AMOUNT
+    const tx = await handleApprove(toApprove)
+
+    if (tx && confirmSwap) {
+      if (getIsTradeApproveResult(tx)) {
+        const approvedAmount = tx.approvedAmount
+        const isApprovedAmountSufficient = Boolean(approvedAmount && approvedAmount >= amountToApproveBig)
+
+        if (isApprovedAmountSufficient) {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/erc20Approve/hooks/useApproveAndSwap.ts:64]: Prevent opening swap confirm if there is not enough allowance after tx mined

File: apps/cowswap-frontend/src/modules/erc20Approve/containers/TradeApproveWithAffectedOrderList/TradeApproveWithAffectedOrderList.tsx [hunk 4]
```diff
@@ -26,17 +27,24 @@ export function TradeApproveWithAffectedOrderList(): ReactNode {
...
-      {showAffectedOrders && partialAmountToApprove && (
-        <ActiveOrdersWithAffectedPermit currency={partialAmountToApprove?.currency} />
+      {isApproveOrPartialPermitRequired && (
+        <>
+          <TradeApproveToggle
+            updateModalState={() => setUserApproveAmountModalState({ isModalOpen: true })}
+            amountToApprove={partialAmountToApprove}
+          />
+          {typeof currentAllowance === 'bigint' && currencyToApprove && (
+            <TradeAllowanceDisplay currentAllowance={currentAllowance} currencyToApprove={currencyToApprove} />
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/erc20Approve/containers/TradeApproveWithAffectedOrderList/TradeApproveWithAffectedOrderList.tsx:46]: Display current allowance amount

File: apps/cowswap-frontend/src/modules/erc20Approve/containers/TradeApproveModal/approveUtils.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  spender: string | undefined
+  currency: Nullish<Currency>
+}
+
+export function processApprovalTransaction(
+  { chainId, currency, account, spender }: ApprovalTransactionParams,
+  txResponse: TransactionReceipt,
+): SetOptimisticAllowanceParams | null {
+  if (txResponse.status !== 1) {
+    throw new Error('Approval transaction failed')
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/erc20Approve/containers/TradeApproveModal/approveUtils.ts:26]: Prevent successful approval path if tx was not successfuly mined

**Review by limitofzero — APPROVED** [review]
Nice work @shoom3301. I've tested with few orders - works well!

**Review by elena-zh — CHANGES_REQUESTED** [review]
Hey @shoom3301 , great job! Thank you!
2 issues/questions; 
1. for permittable tokens 'partial approval' option sends an unlimited amount to approve
image
2. I was not able to test the same flow on the classic eth-flow modals. Could you please add it to the PR? #6351 PR is already merged into develop. 

Thanks!

File: libs/common-utils/src/fractionUtils.ts [hunk 6]
```diff
@@ -179,6 +179,10 @@ export class FractionUtils {
   static simplify(fraction: Fraction): Fraction {
     return reduce(trimZeros(fraction))
   }
+
+  static addPercent<T extends Currency>(amount: CurrencyAmount<T>, percent: Percent): typeof amount {
```
Comment (limitofzero) [inline libs/common-utils/src/fractionUtils.ts:183]: nice helper, thank you!

**Comment by limitofzero** [comment]
@elena-zh could you pls check again?

**Comment by shoom3301** [comment]
@elena-zh I could affect [this fix](https://www.notion.so/cownation/Swap-and-bridge-wrong-button-name-2878da5f04ca8055b102c3937ecd19da), please make sure that it's still fixed
